src/librustc_trans/trans/base.rs

```diff
@@ -2425,6 +2425,14 @@ fn finish_register_fn(ccx: &CrateContext, sp: Span, sym: String, node_id: ast::N
 
 
     if is_entry_fn(ccx.sess(), node_id) {
+        // check for the #[rustc_error] annotation, which forces an
+        // error in trans. This is used to write compile-fail tests
+        // that actually test that compilation succeeds without
+        // reporting an error.
+        if ty::has_attr(ccx.tcx(), local_def(node_id), "rustc_error") {
```

Comment:
I am not clear on why you are limiting use of this attribute to `fn main()`, and having it no-op elsewhere.

I would have had it error *everywhere*, entry or not, especially since the fact that it errors is the basis of the argument for why it does not need a feature gate.